man/texinfo.5
.\" texinfo(5)
.\"
.\" Copyright 1998-2023  Free Software Foundation, Inc.
.\"
.\" Copying and distribution of this file, with or without modification,
.\" are permitted in any medium without royalty provided the copyright
.\" notice and this notice are preserved.
.\"
.de EX
.nf
.ft CW
.in +5
..
.de EE
.in -5
.ft R
.fi
..
.TH TEXINFO 5 "GNU Texinfo" "FSF"
.SH NAME
texinfo \- software documentation system
.SH DESCRIPTION
Texinfo is a documentation system that uses a single source file to
produce both online information and printed output.  It is primarily
designed for writing software manuals.
.PP
For a full description of the Texinfo language and associated tools,
please see the Texinfo manual (written in Texinfo itself).  Most likely,
running this command from your shell:
.EX
info texinfo
.EE
or this key sequence from inside Emacs:
.EX
M-x info RET m texinfo RET
.EE
will get you there.
.SH AVAILABILITY
https://www.gnu.org/software/texinfo/
.br
or any GNU mirror site.
.SH "REPORTING BUGS"
Please send bug reports to bug-texinfo@gnu.org,
general questions and discussion to help-texinfo@gnu.org.
.SH "SEE ALSO"
\fBinfo\fR(1), \fBinstall-info\fR(1), \fBmakeinfo\fR(1), \fBtexi2dvi\fR(1), \fBtexindex\fR(1).
.br
\fBemacs\fR(1), \fBtex\fR(1).
.br
\fBinfo\fR(5).


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 1992-2023 Free Software Foundation, Inc.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

------------------------------------------------------------------------------

* texinfo.tex
 . use @ as the escape character in all index files.  this requires
   new enough texi2dvi (Texinfo 6.7, 2019) for index files to be properly
   processed.
* texi2any
 . info output: output Info-documentlanguage in Local Variables section
   of output file if @documentlanguage is given
 . remove SIMPLE_MENU customization option and tree transformation.
 . remove FRAMES and FRAMESET_DOCTYPE customization options.
 . rename PACKAGE* customization variables to PACKAGE*_OPTION.  For example
   PACKAGE_VERSION customization option is now PACKAGE_VERSION_OPTION.
 . some unused translation files have been removed for the
   "texinfo_document" domain
* info
 . check for init file under ~/.config/texinfo/infokey before checking
  ~/.infokey, in accordance with the XDG Base Directory Specification


7.1 (18 October 2023)
* Language
 . new generic definition commands, @defblock, @defline and @deftypeline,
   for definitions without automatic index entries
 . new @linemacro facility eases use of generic definition commands
 . new command @link creates plain links (supported output formats only)
 . @cartouche takes an argument to specify the cartouche title
 . you can use the new commands @nodedescription and @nodedescriptionblock
   to give text to be used in menu descriptions in Info and HTML output

* texi2any
 . @itemx at the beginning of a @table is now an error, not a warning
 . better validity checking of deeply nested commands
 . check that @set and @clear only appear at the start of a line
 . warn about missing menu entries even if CHECK_NORMAL_MENU_STRUCTURE is
   not set.  you can turn this off by setting CHECK_MISSING_MENU_ENTRY to 0.
 . no longer use --enable-encoding and --disable-encoding to determine
   whether to output encoded characters (instead of entities or commands)
   for HTML, XML, DocBook and LaTeX; instead, use the value of the
   OUTPUT_CHARACTERS customization variable.
 . stricter checks on input encoding, in particular more warnings and
   errors with malformed UTF-8
 . support any input file encoding if support exists in the operating
   system, not just a selected list of encodings
 . resolve an alias referring to another alias at definition time
 . internally, use "source marks" to keep all Texinfo source information that
   is not in the final tree (location of macros, values and included files
   expansion, @if* blocks, DEL comment, and @ protecting end of line on @def*
   lines)
 . HTML output:
    . format @subentry and index entries with @seealso or @seeentry in a more
      similar way to printed output
    . output @shortcontents before @contents by default
    . omit colons after index entries by default.  this can still be
      configured with INDEX_ENTRY_COLON.
    . add @example syntax highlighting as a texi2any extension
    . no more capitalization of @sc argument in HTML Cross-references
    . change @point expansion to U+22C6 in HTML Cross-references
    . if a @node is not associated with a sectioning command but is
      followed by a heading command not usually associated to nodes
      such as @heading and this command appears before other formatted
      content, the heading command is assumed to supply the node heading.
      you can customize this with USE_NEXT_HEADING_FOR_LONE_NODE.
 . Info output:
    . new variable ASCII_DASHES_AND_QUOTES, on by default,
      outputs ASCII characters for literal quote or hyphen characters
      in source, rather than UTF-8.  this makes it easier to search
      Info files.
    . new ASCII_GLYPH variable for using ASCII renditions for glyph
      commands (like @bullet)
    . ASCII_PUNCTUATION still includes the effect of these new variables.
    . new variables AUTO_MENU_DESCRIPTION_ALIGN_COLUMN and AUTO_MENU_MAX_WIDTH
      control the format of descriptions in generated menus
 . XML output:
    . place menu leading text and menu separators in elements instead
      of attributes

* texi2dvi
 . macro expansion with texi2any requires at least version 5.0 (only
   happens with --expand option or with very old texinfo.tex)

* texinfo.tex
 . in @code, ` and ' output by default with backtick and undirected
   single quote glyphs in the typewriter font.  you can still configure
   this using the @codequoteundirected/@codequotebacktick commands.
 . do not insert a space for @ def line continuation, matching the behavior
   of texi2any
 . align section titles in table of contents when more than 10 sections
 . microtype is off by default, for speed
 . page headings generation is no longer linked to the @titlepage command

* info
 . when going Up, position cursor on menu entry for current node
 . allow mouse scrolling support regardless of termcap entries.  this
   supports some more xterm configurations.
 . do not use "/index" as a possible file extension for Info files

* Distribution
  . autoconf 2.71, automake 1.16.5, gettext 0.21



7.0.3 (26 March 2023)
This is a bug-fix release with minimal changes.

* texi2any
  . fix performance regression when Perl binary extension (XS) modules
    are not being used (e.g. with TEXINFO_XS=omit)

* info
  . further fix of recoding of UTF-8 files to ASCII to avoid text
    disappearing from nodes
  . avoid possible freeze at start of a file with `-v nodeline=pointers'


7.0.2 (22 January 2023)
This is a bug-fix release with minimal changes.

* texi2any
  . do not distribute architecture-dependent files
  . build fixed on OpenIndiana 11

* info
  . further fix of recoding of UTF-8 files to ASCII
  . fix check for presence of man pages on Solaris


tp/Texinfo/Convert/Texinfo.pm
# Texinfo.pm: output a Texinfo tree as Texinfo.
#
# Copyright 2010-2023 Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Original author: Patrice Dumas <pertusus@free.fr>
# Parts (also from Patrice Dumas) come from texi2html.pl or texi2html.init.

package Texinfo::Convert::Texinfo;

use 5.00405;
use strict;

# stop \s from matching non-ASCII spaces, etc.  \p{...} can still be
# used to match Unicode character classes.
use if $] >= 5.014, re => '/a';

# To check if there is no erroneous autovivification
#no autovivification qw(fetch delete exists store strict);

use Carp qw(cluck confess);

use Texinfo::Convert::ConvertXS;

# commands definitions
use Texinfo::Commands;
# get_label_element
use Texinfo::Common;

require Exporter;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS);
@ISA = qw(Exporter);

%EXPORT_TAGS = ( 'all' => [ qw(
  convert_to_texinfo
  link_element_to_texi
  target_element_to_texi_label
) ] );

@EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

$VERSION = '7.1dev';

# XS parser and not explicitely unset
my $XS_structuring = ((not defined($ENV{TEXINFO_XS})
                        or $ENV{TEXINFO_XS} ne 'omit')
                       and (not defined($ENV{TEXINFO_XS_PARSER})
                            or $ENV{TEXINFO_XS_PARSER} eq '1')
                       and (not defined($ENV{TEXINFO_XS_STRUCTURE})
                            or $ENV{TEXINFO_XS_STRUCTURE} ne '0'));

my $XS_convert = 0;
$XS_convert = 1 if ($XS_structuring
                    and defined $ENV{TEXINFO_XS_CONVERT}
                    and $ENV{TEXINFO_XS_CONVERT} eq '1');

our $module_loaded = 0;
sub import {
  if (!$module_loaded) {
    if ($XS_convert) {
      # We do not simply override, we must check at runtime
      # that the document tree was stored by the XS parser.
      Texinfo::XSLoader::override(
        "Texinfo::Convert::Texinfo::_convert_tree_with_XS",
        "Texinfo::Convert::ConvertXS::plain_texinfo_convert_tree"
      );
    }
    $module_loaded = 1;
  }
  # The usual import method
  goto &Exporter::import;
}

my %brace_commands           = %Texinfo::Commands::brace_commands;
my %block_commands           = %Texinfo::Commands::block_commands;
my %def_commands             = %Texinfo::Commands::def_commands;

# used in root_heading_command_to_texinfo
my %sectioning_heading_commands = %Texinfo::Commands::sectioning_heading_commands;

my @ignored_types = ('spaces_inserted', 'bracketed_inserted',
'command_as_argument_inserted');
my %ignored_types;
for my $a (@ignored_types) {
  $ignored_types{$a} = 1;
}

# TODO document?
sub link_element_to_texi($)
{
  my $element = shift;
  my $result = '';
  return $result if (!$element->{'extra'});
  if ($element->{'extra'}->{'manual_content'}) {
    $result = '('.convert_to_texinfo($element->{'extra'}->{'manual_content'})
                    .')';
  }
  if ($element->{'extra'}->{'node_content'}) {
    $result .= convert_to_texinfo($element->{'extra'}->{'node_content'});
  }
  return $result
}

# TODO document?
sub target_element_to_texi_label($)
{
  my $element = shift;
  my $label_element = Texinfo::Common::get_label_element($element);
  return convert_to_texinfo({'contents' => $label_element->{'contents'}});
}

# only used in Texinfo::Structuring.  Here and not in Texinfo::Structuring
# to use re => '/a'.
# $REFERENCE_NODE should always be a target element associated to
# a label.
sub check_node_same_texinfo_code($$)
{
  my $reference_node = shift;
  my $node_content = shift;

  my $reference_node_texi;
  if (defined($reference_node->{'extra'}->{'normalized'})) {
    my $label_element = Texinfo::Common::get_label_element($reference_node);
    $reference_node_texi = convert_to_texinfo(
                                {'contents' => $label_element->{'contents'}});
    $reference_node_texi =~ s/\s+/ /g;
  } else {
    $reference_node_texi = '';
  }

  my $node_texi;
  if ($node_content) {
    my $contents_node = $node_content;
    if ($node_content->{'contents'}->[-1]->{'type'}
        and $node_content->{'contents'}->[-1]->{'type'}
                                      eq 'space_at_end_menu_node') {
      $contents_node = {'contents' => [@{$node_content->{'contents'}}]};
      pop @{$contents_node->{'contents'}};
    }
    $node_texi = convert_to_texinfo($contents_node);
    $node_texi =~ s/\s+/ /g;
  } else {
    $node_texi = '';
  }
  return ($reference_node_texi eq $node_texi);
}

# for debugging.
sub root_heading_command_to_texinfo($)
{
  my $element = shift;
  my $tree;
  if ($element->{'cmdname'}) {
    if (($element->{'cmdname'} eq 'node'
         or $sectioning_heading_commands{$element->{'cmdname'}})
        and $element->{'args'}->[0]->{'contents'}) {
      $tree = $element->{'args'}->[0]->{'contents'};
    }
  } else {
    return "Not a command";
  }
  if ($tree) {
    return '@'.$element->{'cmdname'}.' '
                .convert_to_texinfo({'contents' => $tree});
  } else {
    return '@'.$element->{'cmdname'};
  }
}

# Following subroutines deal with transforming a texinfo tree into texinfo
# text.  Should give the text that was used parsed, except for a few cases.

# This is used if the document is available for XS, but XS is not
# used (most likely $TEXINFO_XS_CONVERT is 0).
sub _convert_tree_with_XS($)
{
  my $root = shift;

  return _convert_to_texinfo($root);
}

# expand a tree to the corresponding texinfo.
sub convert_to_texinfo($);
sub convert_to_texinfo($)
{
  my $element = shift;

  if (defined($element->{'tree_document_descriptor'})) {
    return _convert_tree_with_XS($element);
  }
  return _convert_to_texinfo($element);
}

sub _convert_to_texinfo($);
sub _convert_to_texinfo($)
{
  my $element = shift;

  confess "convert_to_texinfo: element undef" if (!defined($element));
  confess "convert_to_texinfo: bad element type (".ref($element).") $element"
     if (ref($element) ne 'HASH');
  my $result = '';

  return '' if ($element->{'type'}
                and ($ignored_types{$element->{'type'}}));
  if (defined($element->{'text'})) {
    $result .= $element->{'text'};
  } else {
    if ($element->{'cmdname'}
        or ($element->{'type'} and $element->{'type'} eq 'def_line')) {
      $result .= _expand_cmd_args_to_texi($element);
    } else {
      if ($element->{'type'}
          and ($element->{'type'} eq 'bracketed_arg'
               or $element->{'type'} eq 'bracketed_linemacro_arg')) {
        $result .= '{';
      }
      if ($element->{'info'}
          and $element->{'info'}->{'spaces_before_argument'}) {
        $result .= $element->{'info'}->{'spaces_before_argument'}->{'text'};
      }
    }
    if (defined($element->{'contents'})) {
      foreach my $child (@{$element->{'contents'}}) {
        $result .= _convert_to_texinfo($child);
      }
    }
    if ($element->{'info'} and $element->{'info'}->{'spaces_after_argument'}) {
      $result .= $element->{'info'}->{'spaces_after_argument'}->{'text'};
    }
    if ($element->{'info'} and $element->{'info'}->{'comment_at_end'}) {
      $result .= _convert_to_texinfo($element->{'info'}->{'comment_at_end'})
    }
    $result .= '}' if ($element->{'type'}
                       and ($element->{'type'} eq 'bracketed_arg'
                            or $element->{'type'} eq 'bracketed_linemacro_arg'));
  }
  return $result;
}

# expand a command argument as texinfo.
sub _expand_cmd_args_to_texi($) {
  my $cmd = shift;

  my $cmdname = $cmd->{'cmdname'};
  $cmdname = '' if (! defined($cmd->{'cmdname'}));
  my $result = '';

  if ($cmdname) {
    $result = '@'.$cmdname;

    # this is done here otherwise for some constructs, there are
    # no 'args', and so the space is never readded.
    if ($cmd->{'info'}
        and $cmd->{'info'}->{'spaces_after_cmd_before_arg'}) {
      $result .= $cmd->{'info'}->{'spaces_after_cmd_before_arg'}->{'text'};
    }
  }
  # arg_line set for line_commands with type special and @macro
  if ($cmd->{'info'} and defined($cmd->{'info'}->{'arg_line'})) {
    $result .= $cmd->{'info'}->{'spaces_before_argument'}->{'text'}
      if $cmd->{'info'} and $cmd->{'info'}->{'spaces_before_argument'};
    $result .= $cmd->{'info'}->{'arg_line'};
  } elsif ($cmd->{'args'}) {
    my $braces;
    $braces = 1 if (scalar(@{$cmd->{'args'}})
                    and ($cmd->{'args'}->[0]->{'type'}
                          and ($cmd->{'args'}->[0]->{'type'} eq 'brace_command_arg'
                               or $cmd->{'args'}->[0]->{'type'} eq 'brace_command_context')));
    $result .= '{' if ($braces);
    if ($cmdname eq 'verb') {
      $result .= $cmd->{'info'}->{'delimiter'};
    }
    $result .= $cmd->{'info'}->{'spaces_before_argument'}->{'text'}
       if ($cmd->{'info'} and $cmd->{'info'}->{'spaces_before_argument'});
    my $with_commas = 0;
    if (($block_commands{$cmdname}
         # block line commands with arguments not separated by commas
         and not ($def_commands{$cmdname}
                  or $block_commands{$cmdname} eq 'multitable'))
        or $cmdname eq 'node'
        or exists($brace_commands{$cmdname})
        or ($cmd->{'type'} and $cmd->{'type'} eq 'definfoenclose_command')) {
      $with_commas = 1;
    }
    my $arg_nr = 0;
    foreach my $arg (@{$cmd->{'args'}}) {
      next if $arg->{'type'} and $ignored_types{$arg->{'type'}};
      if ($with_commas) {
        $result .= ',' if ($arg_nr);
        $arg_nr++;
      }
      $result .= _convert_to_texinfo($arg);
    }
    if ($cmdname eq 'verb') {
      $result .= $cmd->{'info'}->{'delimiter'};
    }
    $result .= '}' if ($braces);
  } else {
    $result .= $cmd->{'info'}->{'spaces_before_argument'}->{'text'}
      if $cmd->{'info'} and $cmd->{'info'}->{'spaces_before_argument'};
  }
  return $result;
}

1;
__END__

=head1 NAME

Texinfo::Convert::Texinfo - Convert a Texinfo tree to Texinfo code

=head1 SYNOPSIS

  use Texinfo::Convert::Texinfo qw(convert_to_texinfo);

  my $texinfo_text = convert_to_texinfo($tree);

=head1 NOTES

The Texinfo Perl module main purpose is to be used in C<texi2any> to convert
Texinfo to other formats.  There is no promise of API stability.

=head1 DESCRIPTION

C<Texinfo::Convert::Texinfo> converts a Texinfo tree (described in
L<Texinfo::Parser>) to Texinfo code.  If the Texinfo tree results from
parsing some Texinfo document, The converted Texinfo code should be
exactly the same as the initial document, except that user defined @-macros
and C<@value> are expanded, and some invalid code is discarded.

=head1 METHODS

=over

=item $texinfo_text = convert_to_texinfo($tree)
X<C<convert_to_texinfo>>

Converts the Texinfo tree I<$tree> to Texinfo code.

=back

=head1 AUTHOR

Patrice Dumas, E<lt>pertusus@free.frE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2010- Free Software Foundation, Inc.  See the source file for
all copyright years.

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or (at
your option) any later version.

=cut
